### ISSUE TITLE
Request: A DeleteBehavior that does absolutely nothing on delete and on SaveChanges

### ISSUE THREAD

**Issue by Wain123** [aspnet/EntityFramework#8633]
DeleteBehavior.Restrict behaves the same as DeleteBehavior.SetNull for tracked entities, details are in #8342 

My problem is that, if I delete an entity and forget to handle its children, EF will simply set their foreign keys to null and let me save the mistake to the DB. It goes out of its way to make sure I don't notice my mistake.

I would like to have a DeleteBehavior option that simply does nothing. If a navigation property of a child entity points to a deleted parent entity, I want SaveChanges to send a DELETE command for the parent and leave the child alone, so that my DB can give me a foreign key constraint exception. Alternatively, EF could also give me an exception on SaveChanges telling me that something points to a deleted entity.

Whether the FK is nullable or not also shouldn't matter, don't null it even if it's nullable.

**Comment by ajcvickers** [comment]
@Wain123 We discussed this at length and came to a couple of conclusions:
* This API is really misleading. We had decided to better document it (see #8342), but after further discussion we are now going to obsolete it and go back to an API that just specified cascade delete behavior. This will mean there will no longer any confusion over setting SetNull or Restrict and having the DbContext tracking continue to behave in the same way.
* The behavior in the context is driven by the model metadata. For required relationships, cascade deletes are switched on by default, but if switched off, then EF will throw if an FK needs to be nulled but cannot be. For optional relationships, cascades deletes are off by default (but can be switched on) and the FK is nulled because this is the best semantic representation of what happened in the model. We're not going to support optional relationships where we don't attempt to null FKs. This would leave the model in an inconsistent state, and one of the functions of an OR/M is to help the model and relationships stay in a consistent state.
